### ISSUE TITLE
Adding evaluation to the sft trainer

### ISSUE THREAD

**Issue by GiilDe** [2U1/Qwen-VL-Series-Finetune#230]
It could very helpful if you'd mention what are the steps needed in order to add evaluation to the sft trainer. I added `--eval_path` and `--eval_image_folder` and also passed my metric callable, i.e. `QwenSFTTrainer(...compute_metrics=metric_func...)`. The issue is that the when I decode the logits passed to my metric function it seems that it contains lengthy gibberish. I suspect that it is because inside the `prediction_step` function - https://github.com/2U1/Qwen-VL-Series-Finetune/blob/58a3e2bb59be6f39da0bfee7fd34e55ea432b0e5/src/trainer/sft_trainer.py#L210 the call is `model(**inputs)` and not `model.generate(**inputs)`. So I changed to use `model.generate(**inputs)`, but then the issue was that the `inputs_ids` contain not only the quesiton but also the answer, which made the model return empty answers. So I removed the concatenation of the gpt response to the question in the validation dataset. This seemingly worked, but unfortunately, it seems now that the answers I receive are very low-quality, I'm not sure why. Anyhow, can you explain how to add string-level validation to the training in this code? it can be very helpful as it seems not to be so trivial

**Comment by 2U1** [comment]
At the moment the SFT trainer in this repo only does token-level
evaluation during training. This means that the tensor you see inside
`compute_metrics` is the full logit tensor of shape
(batch, seq_len, vocab_size), not token ids. Decoding that directly with
the tokenizer will always look like “gibberish”.

Instead of changing `prediction_step` to use `model.generate`, you should
use the standard `preprocess_logits_for_metrics` hook from the HF
Trainer API and turn the logits into token ids before they are passed to
`compute_metrics`. For example:
```
    def preprocess_logits_for_metrics(logits, labels):
        if isinstance(logits, tuple):
            logits = logits[0]
        return logits.argmax(dim=-1)

    def compute_metrics(eval_preds):
        preds = eval_preds.predictions      # (batch, seq_len) of token ids
        labels = eval_preds.label_ids       # (batch, seq_len) of token ids / -100

        labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
        pred_texts  = tokenizer.batch_decode(preds, skip_special_tokens=True)
        label_texts = tokenizer.batch_decode(labels, skip_special_tokens=True)

        # compute your string-level metric here (exact match, BLEU, etc.)
        ...

    trainer = QwenSFTTrainer(
        ...,
        compute_metrics=compute_metrics,
        preprocess_logits_for_metrics=preprocess_logits_for_metrics,
    )
```

You can post-process the decoded strings to extract
only the last assistant message (e.g. using a special `<answer>` token
or your chat template) before computing the metric.

Supporting true generation-based evaluation (using `model.generate` in
`prediction_step`) would require a separate eval dataset where
`input_ids` only contain the prompt and `labels` only contain the
answer, so for now I’d recommend sticking to the
`preprocess_logits_for_metrics` + string post-processing approach.

**Comment by bbeomdev** [comment]
@GiilDe Maybe.. it might help to check this out.

https://github.com/2U1/Qwen-VL-Series-Finetune/pull/183

**Comment by 2U1** [comment]
@GiilDe @bbeomdev   

Sorry I'm late, I was bit busy. I've implemented generation-based evaluation support for the SFT trainer.

  ### Root Cause

  The issue you encountered was due to:

  1. **`prediction_step` using `model(**inputs)`** - This returns logits, not generated text. Decoding logits directly produces gibberish.

  2. **When switching to `model.generate()`** - The `input_ids` already contain both the question AND the answer (concatenated during data preprocessing), causing the model to return empty outputs since there's nothing left to generate.

  3. **Empty output with long prompts** - Vision tokens expand the actual sequence length beyond `input_ids.shape[1]`, so trimming by `len(input_ids)` doesn't work correctly.

  ### Solution

  I've overridden `evaluation_loop` in `QwenSFTTrainer` to:

  1. Extract prompt-only tokens using the `labels` mask (where `labels == -100` indicates prompt tokens)
  2. Use `model.generate()` with the extracted prompts
  3. Pass generated text strings to `compute_metrics`

  ### Usage

  ```python
  from src.trainer import QwenSFTTrainer, GenerativeEvalPrediction

  def compute_metrics(eval_pred: GenerativeEvalPrediction):
      predictions = eval_pred.predictions  # List[str]
      references = eval_pred.references    # List[str]

      correct = sum(1 for p, r in zip(predictions, references) if p.strip() == r.strip())
      return {"accuracy": correct / len(predictions)}

  trainer = QwenSFTTrainer(
      model=model,
      processing_class=processor,
      args=training_args,
      compute_metrics=compute_metrics,
      **data_module
  )
```
  Training arguments:
  --eval_path /path/to/eval.json \
  --eval_strategy steps \
  --eval_steps 500 \
  --generation_max_new_tokens 256 \
  --prediction_loss_only False

  I've also added detailed documentation in the README under "Evaluation during Training" section with step-by-step instructions and examples.

  Let me know if you have any questions!

**Comment by GiilDe** [comment]
@2U1 Thank you!!!
